py/vm.c

```diff
@@ -850,8 +850,26 @@ unwind_jump:;
                         step = POP();
                     }
                     mp_obj_t stop = POP();
-                    mp_obj_t start = TOP();
-                    SET_TOP(mp_obj_new_slice(start, stop, step));
+                    mp_obj_t start = POP();
+                    if ((*ip == MP_BC_LOAD_SUBSCR || *ip == MP_BC_STORE_SUBSCR) && mp_obj_is_native_type(mp_obj_get_type(TOP()))) {
+                        // Fast path optimisation for when the BUILD_SLICE is
+                        // immediately followed by a LOAD/STORE_SUBSCR for a
+                        // native type to avoid needing to allocate the slice
+                        // on the heap. In some cases (e.g. a[1:3] = x) this
+                        // can result in no allocations at all. We can't do
+                        // this for instance types because the get/set/delattr
+                        // implementation may keep a reference to the slice.
+                        byte op = *ip++;
+                        mp_obj_slice_t slice = { .base = { .type = &mp_type_slice }, .start = start, .stop = stop, .step = step };
```

Comment:
This possibly increases C stack usage of this function.  Will need to quantify that.

Reply:
Indeed... +24 bytes of stack.

I've added a commit to address this (using alloca). Slight code size increase though (was +120, now TODO).

Reply:
Reverted this for now... too complicated to figure out how this works with `MICROPY_ENABLE_PYSTACK`.